flight/PiOS/STM32F4xx/pios_dmashot.c

```diff
@@ -0,0 +1,591 @@
+	if (s_timer->dma->master_timer) {
+		// The DMABase shift is in count of 32-bit registers. 16-bit registers are padded with 16-bit reserved ones,
+		// and thus defacto 32-bit.
+		dma.DMA_PeripheralBaseAddr = ((uint32_t)&s_timer->dma->timer->CR1) + ((s_timer->dma->master_config & 0x00FF) << 2);
+	} else {
+		dma.DMA_PeripheralBaseAddr = (uint32_t)(&s_timer->dma->timer->DMAR);
+	}
+
+	if (PIOS_DMAShot_HalfWord(s_timer)) {
+		dma.DMA_MemoryDataSize = DMA_MemoryDataSize_HalfWord;
+		dma.DMA_PeripheralDataSize = DMA_PeripheralDataSize_HalfWord;
+	} else {
+		dma.DMA_MemoryDataSize = DMA_MemoryDataSize_Word;
+		dma.DMA_PeripheralDataSize = DMA_PeripheralDataSize_Word;
+	}
+
+	dma.DMA_MemoryInc = DMA_MemoryInc_Enable;
+	dma.DMA_PeripheralInc = DMA_PeripheralInc_Disable;
+
+	dma.DMA_DIR = DMA_DIR_MemoryToPeripheral;
+	dma.DMA_Mode = DMA_Mode_Normal;
+
+	dma.DMA_BufferSize = PIOS_DMAShot_GetNumChannels(s_timer) * DMASHOT_STM32_BUFFER;
+
+	dma.DMA_Priority = DMA_Priority_VeryHigh;
+	dma.DMA_MemoryBurst = DMA_MemoryBurst_Single;
+	dma.DMA_PeripheralBurst = DMA_PeripheralBurst_Single;
+
+	dma.DMA_FIFOMode = DMA_FIFOMode_Enable;
+	dma.DMA_FIFOThreshold = PIOS_DMAShot_GetFIFOCadence(PIOS_DMAShot_GetNumChannels(s_timer));
+
+	DMA_Init(s_timer->dma->stream, &dma);
+
+	// Don't do interrupts.
+	DMA_ITConfig(s_timer->dma->stream, DMA_IT_TC, DISABLE);
+}
+
+/**
+ * @brief Initializes and configures the known DMA channels for DMAShot operation.
+ */
+void PIOS_DMAShot_InitializeDMAs()
+{
+	// If there's nothing setup, fail hard. We shouldn't be getting here.
+	PIOS_Assert(dmashot_cfg && servo_timers);
+
+	for (int i = 0; i < MAX_TIMERS; i++) {
+		struct servo_timer *s_timer = servo_timers[i];
+		if (!s_timer || !s_timer->sysclock)
+			continue;
+
+		if (!s_timer->buffer.ptr) {
+			// Allocate buffer at timer resolution.
+			if (PIOS_DMAShot_HalfWord(s_timer)) {
+				s_timer->buffer.ptr =
+						PIOS_malloc(PIOS_DMAShot_GetNumChannels(s_timer) * DMASHOT_STM32_BUFFER * sizeof(uint16_t));
+				memset(s_timer->buffer.ptr, 0, PIOS_DMAShot_GetNumChannels(s_timer) * DMASHOT_STM32_BUFFER * sizeof(uint16_t));
+			} else {
+				s_timer->buffer.ptr =
+						PIOS_malloc(PIOS_DMAShot_GetNumChannels(s_timer) * DMASHOT_STM32_BUFFER * sizeof(uint32_t));
+				memset(s_timer->buffer.ptr, 0, PIOS_DMAShot_GetNumChannels(s_timer) * DMASHOT_STM32_BUFFER * sizeof(uint32_t));
+			}
+		}
+
+		PIOS_DMAShot_DMASetup(s_timer);
+	}
+}
+
+/**
+ * @brief Triggers the configured DMA channels to fire and send throttle values to the timer DMAR and optional CCRx registers.
+ */
+void PIOS_DMAShot_TriggerUpdate()
+{
+	// If there's nothing setup, fail hard. We shouldn't be getting here.
+	PIOS_Assert(dmashot_cfg && servo_timers);
+
+	for (int i = 0; i < MAX_TIMERS; i++) {
+		struct servo_timer *s_timer = servo_timers[i];
+		if (!s_timer || !s_timer->sysclock)
+			continue;
+
+		DMA_Cmd(s_timer->dma->stream, DISABLE);
+		while (DMA_GetCmdStatus(s_timer->dma->stream) == ENABLE) ;
```

Comment:
I suggest we wait for completion --before disabling--, so that if jitter causes us to show up when another update is in progress we wait for it to be done.